docs: add RFD for linux job worker service

dustinspecker opened the pull request:
This RFD lays out how to design a Linux job worker service for running arbitrary Linux processes with some degree of isolation and resource limits.

Users should be able to interact with the service via a CLI to start jobs, stop jobs, check status of jobs, and stream a job's output.

sclevine commented on rfd/0001-linux-job-worker-service.md:
```diff
@@ -0,0 +1,408 @@
+
+This new type can store all process output in a `[]byte`. Consider a mutex to handle concurrent read/writes.
+
+The Job's `Stream` function will invoke this new type's `Stream` function. A new goroutine should be created per client that reads from the new type. The goroutine will track the last byte read and send all new bytes to the client. The goroutine should finish when
```
While I think this implementation will work, it actively pushes bytes to the consumer (via channel) that might not be needed immediately. Remember that you have as much memory as you need -- even `( len(stdout) + len(stderr) ) x # of clients` is okay.

In general though, the idea of creating an in-memory structure that broadcasts via sync.Cond sounds ideal 🙂 

Think carefully about how you can efficiently fan-out the writes, and how you can improve the API

dustinspecker commented on rfd/0001-linux-job-worker-service.md:
````diff
@@ -0,0 +1,408 @@
+```go
+type JobConfig struct {
+  JobExecutorPath  string
+  CPU              float64
+  MemBytes         int64
+  IOBytesPerSecond int64
````
Sorry, I completely missed describing those. Added. And also added how these limits would be used in the cgroup.

dustinspecker commented on rfd/0001-linux-job-worker-service.md:
```diff
@@ -0,0 +1,408 @@
+> server binary instead of the server binary and the `JobExecutorPath` binary. The server could provide `/proc/self/exe` to the library
+> as `JobExecutorPath`.
+
+The `JobExecutorPath` may seem odd, but this enables a "hook" to add the new process's PID to the cgroup before actually executing
```
Thank you for asking this because I realized I overlooked `UseCgroupFD` and `CgroupdFD` added (fully implemented?) in Go 1.22: https://github.com/golang/go/issues/51246 !!!

I need  to do some more research, but I think these options on `exec.Cmd`'s [https://pkg.go.dev/syscall#SysProcAttr](https://pkg.go.dev/syscall#SysProcAttr) can prevent the `JobExecutorPath` binary from having to do anything with cgroups directly. The library can 100% handle that part.

My limited knowledge is leaning towards the `JobExecutorPath` binary would still need to mount a new proc filesystem before executing the user's comand. Mounting a new proc filesystem could definitely be another function in the library. I've made a note of that.

Also, sorry for this doc now hinting towards `UseCgroupFD` and sort of describing the previous proposed plan with the `JobExecutorPath` appending the PID itself before executing. I want to research more, but for now this RFD kind of handles both. I'll poke around some more, but I think I'll end up being able to get rid of `JobExecutorPath` binary from messing with groups altogether.

dustinspecker replied:
I also need to think of a better way to handle the device ID in `io.max` (or give a way to hardcode it). I haven't used `io.max` before. I was hoping there was an _apply this limit to all devices_ config.

sclevine replied:
You'll need the re-exec logic regardless for (at least) /proc mounting, so either way is great. Just make sure the RFD is consistent 🙂

dustinspecker replied:
Updated RFD so that it only uses `exec.Cmd`'s `UseCgroupFD` and `CgroupFD`.

dustinspecker commented on rfd/0001-linux-job-worker-service.md:
```diff
@@ -0,0 +1,408 @@
+`Start` will also create a new cgroup for the job and set the CPU, memory, and IO limits. It will be up to the `jobExecutorPath` to append the PID to the cgroup's `cgroup.procs` file as described above.
+
+The configured `exec.Cmd` will execute `jobExecutorPath`. It is expected that `jobExecutorPath` is a binary that will do the following:
+
```
Sorry, I'm not  following your comment. I've tried to re-word/organize this. It's not obvious to me what I'm missing here.

My understanding (very limited to be honest) is that when `JobExecutorPath` running within a new mount namespace mounts a new proc filesystem that it is not reflected on the host's mount namespace. Yep, in the current design, if the user wants to run `ps -ef` then they would see `JobExecutorPath` as PID 1 and `ps -ef`'s PID. But no other processes will be displayed.

sclevine replied:
> that it is not reflected on the host's mount namespace

I recommend testing out mounting /proc from a new mount namespace. Look at what happens to /proc on the host.

dustinspecker replied:
Oops, I think I see what you're saying now. I added some clarity about specifying an unshare flag for mount namespace so new mounts don't show up within the host's mount namespace. Also, added clarity that the `JobExecutorPath` should create a new directory to mount a new proc filesystem to.

Please let me know if that's what you were referring to. Sorry for the confusion - I definitely wasn't clear here.


If you're referring to another step that I'm missing here then I definitely do not know what I'm missing.

sclevine replied:
You need to mount `/` in a private subtree within the namespace first, in order to prevent the /proc mount from applying in the host namespace. `/` is often mounted as a shared subtree.

https://docs.kernel.org/filesystems/sharedsubtree.html

dustinspecker replied:
I'm definitely misunderstanding something.

If `Start()` runs a `JobExecutorPath` with the following flags:

```go
cmd := exec.Command(JobExecutorPath, arguments...)
cmd.SysProcAttr = &syscall.SysProcAttr{
    Cloneflags:   syscall.CLONE_NEWPID | syscall.CLONE_NEWNS | syscall.CLONE_NEWNET,
    Unshareflags: syscall.CLONE_NEWNS,
}
```

and then `JobExecutorPath` mounts a new proc filesystem (`syscall.Mount("proc", "/proc", "proc", 0)`) then I'm not seeing any issues with the host's mounts or proc file system.


If I drop the `Unshareflags` then when `JobExecutorPath` mounts a new proc filesystem then I can confirm the host's proc filesystem is messed up. If the `Unshareflags` isn't provided then doing what you said (`JobExecutorPath` mounting `/` with `syscall.MS_PRIVATE|syscall.MS_REC`) does prevent messing up the host's proc filesystem as well.


Is there something else I'm missing here? I'm cool adding the mounting of `/` with `MS_PRIVATE|MS_REC` flags, but I don't understand what it's doing when the `Unshareflags` is specified.

sclevine replied:
Interesting! Last time I tested this, mounting / as private was necessary to escape the shared subtree. Could be distro-specific (I think systemd on some distros will make / private by default), or maybe this has always worked with `unshare()`. All good if you're confident it's working 🙂

dustinspecker replied:
You got me curious. Running `findmnt -o TARGET,PROPAGATION` in the host mount namespace shows `/` is mounted as shared.

This is also true when running in the new mount namespace without specifying the unshare flag. If the unshare flag for `CLONE_NEWNS` is provided then running `findmnt -o TARGET,PROPAGATION` within the new mount namespace shows `/` is mounted as private.

I'll try to look around for some documentation to back this up, mostly for my own curiosity.

dustinspecker commented on rfd/0001-linux-job-worker-service.md:
```diff
@@ -0,0 +1,408 @@
+
+`Status` returns the current status of the job. Refer to [JobStatus](#type-jobstatus) for more information.
+
+##### func (*Job) Stream(context.Context) <-chan []byte
```
The goal of the context was to cancel/stop receiving data.

I need to think on this. I will definitely admit an API using a context and returning a channel feels _very_ clunky/unintuitive.

sclevine replied:
> an API using a context and returning a channel feels very clunky/unintuitive

Exactly. Just focusing on the API design, how would you usually stream bytes in Go?

dustinspecker replied:
I've updated the docs for `Stream` to no longer take any parameters and now returns an `OutputReader` that implements `io.Reader` (and ultimately is able to read a process's output from another type, `Output`). I've added more information about `OutputReader` and `Output` towards the bottom of the RFD.

Thank you for calling this out! This is a way better design that doesn't require the user to close or cancel anything.

sclevine replied:
Much better. I might lean towards returning io.Reader or io.ReadCloser directly instead of a custom type, but either way works. A custom type is more open to extension in the future.

dustinspecker commented on rfd/0001-linux-job-worker-service.md:
```diff
@@ -0,0 +1,408 @@
+
+- append the PID to the provided `cgroup.procs` file
+- mount a new proc filesystem
+- execute the provided `command` and `arguments`
```
I think I've added some clarity around that.

I currently have both of these being fork + exec via [exec.Cmd](https://pkg.go.dev/os/exec#Cmd), so that means `JobExecutorPath` ends up being PID 1 in the new PID namespace. My understanding is that when PID 1 is killed  the other processes are killed. And if `JobExecutorPath` is waiting for the one command to run, then it'll terminate after the command finishes. And if a process becomes orphaned, then it'll become a child of PID 1 in the PID namespace.

If a command spawns child processes and quits immediately, then this will kill all processes in the PID namespace to my understanding.

sclevine replied:
What signal will Stop send, and to which PID or PIDs?

> My understanding is that when PID 1 is killed the other processes are killed. And if JobExecutorPath is waiting for the one command to run, then it'll terminate after the command finishes. And if a process becomes orphaned, then it'll become a child of PID 1 in the PID namespace.

These are valid assumptions, and fork+exec for both gives you a ton of flexibility (similar to a process manager) for supporting graceful termination of both long-running processes and short-running tasks.

> If a command spawns child processes and quits immediately, then this will kill all processes in the PID namespace to my understanding.

Not sure I understand this part. If the executer is PID 1, the process in PID 2, and the child is PID 3, terminating PID 2 may not necessarily terminate PID 3.

dustinspecker replied:
> What signal will Stop send, and to which PID or PIDs?

For now, my intention is for Stop to use [exec.Cmd.Process.Kill](https://pkg.go.dev/os#Process.Kill). Since the `JobExecutorPath` is PID 1 in the new namespace this will kill all other processes in the namespace. Like you mentioned, a future improvement could be enabling graceful terminations by first sending a SIGTERM via [exec.Cmd.Process.Signal](https://pkg.go.dev/os#Process.Signal) to PID 1. Then PID 1 could handle SIGTERM and also send SIGTERM to all of its children processes.  I'll add these to the RFD.

>> If a command spawns child processes and quits immediately, then this will kill all processes in the PID namespace to my understanding.

> Not sure I understand this part. If the executer is PID 1, the process in PID 2, and the child is PID 3, terminating PID 2 may not necessarily terminate PID 3.


I worded that poorly, sorry.

If a command (PID 2) spawns child processes (PID 3) and then that command  (PID 2) quits immediately and leaves orphan processes (PID 3) around, then the child processes become child processes of PID 1. Then later when PID 1 is killed then these child processes (PID 3) are also killed.

tcsc commented on rfd/0001-linux-job-worker-service.md:
```diff
@@ -0,0 +1,439 @@
+1. Demonstrate starting, checking the job's status, and stopping a job.
+1. Demonstrate starting a job, streaming the job output, and validating the stream completes when the process completes.
+1. Demonstrate a user may not interact with another user's job.
+1. Demonstrate a job may not make network requests.
```
I'd also like to see test cases asserting that the `cgroup` limits are being enforced. Due to the nature of this exercise I'm not expecting that all or them would be practical to implement, but a quick discussion of how you'd go about it would be useful.

dustinspecker replied:
Great suggestion! I added some example test scenarios, such as using `dd` to verify io.max settings and `sha1sum` to verify CPU.max settings.

tcsc commented on rfd/0001-linux-job-worker-service.md:
```diff
@@ -0,0 +1,439 @@
+
+#### Authorization
+
+The API treats the client's common name (CN) in the certificate as the user's identity.
```
Can you add a couple of lines about how the implementation will get access to the CN from the client cert?

dustinspecker replied:
Good call! I added a couple of lines about being able to extract the CN from the certificate from the context in each handler.

tcsc commented on rfd/0001-linux-job-worker-service.md:
```diff
@@ -0,0 +1,439 @@
+}
+
+message Job {
+  string id = 1;
```
How will Job IDs be generated?

dustinspecker replied:
I added a tiny bit about how the server will work at a high level. When it's told to start a new job, the server generates a new UUID v4 as the ID and returns it to the client. It'll also use this ID as a key for storing jobs in a map.

tcsc commented on rfd/0001-linux-job-worker-service.md:
```diff
@@ -0,0 +1,439 @@
+
+##### func (*Job) Stop() error
+
+`Stop` will kill the process.
```
How will the job processes be killed? Will the jobs be given a grace period to finish up once signalled?

dustinspecker replied:
I added a note that the process will be killed via [kill](https://pkg.go.dev/os#Process.Kill). Since JobExecutorPath is PID 1 in the new pid namespace when it terminates all other processes running in the new pid namespace are also killed.

I also added a note that graceful termination could be handled in the future by sending SIGTERM to JobExuectorPath, which could handle SIGTERM by sending a SIGTERM signal to the child process (the user's desired command). If the process has not terminated in some amount of time then SIGKILL is used.

Please let me know if that's cool with you or if you would like to see graceful termination supported in this.

sclevine replied:
Having `Stop()` send SIGKILL seems unusual, especially since your RFD calls out support for long-running processes explicitly. Many commands are not intended to be killed by the kernel without any notice.

It's definitely important not to go too far out of scope for the challenge, but we expect the job lifecycle management to be relatively robust 🙂

dustinspecker replied:
Sounds good. I'll update the RFD so that the library's `Stop` sends `SIGTERM` to the `JobExecutorPath` (and appropriately send `SIGTERM` to child process), wait some amount of time (probably something like 10 seconds as a hardcoded value for this), check if PID 1 is alive, and if it is then send `SIGKILL`.

sclevine replied:
This sounds great! And of course, think about how you'll deal with child processes.

Remember that you're providing the logic for JobExecutorPath, so if it receives SIGTERM, you'll want to specify how it handles that signal. If it quits immediately, all processes will get SIGKILL.

dustinspecker replied:
I updated the RFD to handle graceful shutdowns.

JobExecutorPath is now expected to handle SIGTERM. When it receives SIGTERM, it will send SIGTERM to the child process (user's command). JobExecutorPath must not terminate until the child process does.

On the library side, `Stop` will send SIGTERM to JobExecutorPath. If after 10 seconds, JobExecutorPath is alive then `Stop` will send SIGKILL to JobExecutorPath.

sclevine replied:
Sounds great. Note that you may have multiple processes under JobExecutorPath, and gracefully terminating all of them may not be as simple as cmd.Wait()

tcsc reviewed (commented):
Looks good to me in general - just a few questions in the comments.

The only major thing I can think of is that I'd like to see a brief discussion of what the glue between the GRPC API and the `Job` library will look like. Things like how the server will track the running/finished jobs, handle concurrent requests and suchlike.

dustinspecker commented:
> Looks good to me in general - just a few questions in the comments.
> 
> The only major thing I can think of is that I'd like to see a brief discussion of what the glue between the GRPC API and the `Job` library will look like. Things like how the server application will
> 
> * track the running/finished jobs,
> * extract the users CN from the client certificate and use that for authorization
> * handle concurrent job start requests and suchlike.

I've added a few lines on how the server works.

- server can store jobs in a map, preferably a [sync.Map](https://pkg.go.dev/sync#Map)
- I've added information about how the CN can be extracted in each handler via the context.
- I've added a couple of notes about gRPC handlers each running in a new goroutine and the "job store" needs to be safe for concurrent use. I also added a note in the library that the Job struct should be safe for concurrent changes as well.

Thank you so much for the feedback!

tcsc commented on rfd/0001-linux-job-worker-service.md:
```diff
@@ -0,0 +1,510 @@
+> server binary instead of the server binary and the `JobExecutorPath` binary. The server could provide `/proc/self/exe` to the library
+> as `JobExecutorPath`.
+
+The `JobExecutorPath` may seem odd, but this enables a "hook" to create and mount a new proc filesystem before forking + executing the desired command with any arguments.
```
Can you add in a sentence as to why this is a useful behaviour?

```suggestion
The `JobExecutorPath` may seem odd, but this enables a "hook" to create and mount a new proc filesystem before forking + executing the desired command with any arguments. This is important because...
```

dustinspecker replied:
I've added a sentence. Please let me know if that's clearer.

tcsc commented on rfd/0001-linux-job-worker-service.md:
```diff
@@ -0,0 +1,510 @@
+It is expected that `JobExecutorPath` will mount a new proc filesystem and then fork+execute the user's command with any arguments.
+
+> Note: `JobExecutorPath` may be a new binary entirely created from this design for the library to use. Eventually, it may make more
+> sense for the server binary to be a compatible `JobExecutorPath` that the library uses. This way to deploy the server requires only the one
```
Is your intent here your `job` library will supply an implementation for this behaviour that a server binary can use?

dustinspecker replied:
I've added clarity on this. Please let me know your thoughts.

So, phase 1 of the library only will have a separate binary for `JobExecutorPath`.
In phase 2, when the server binary is being created, the server binary will also satisfy the `JobExecutorPath` requirements, so the first binary will go away. The library will be told the path of the `JobExecutorPath` in both situations, so no changes to the library is required for the server to be a `JobExecutorPath`.

And the library will have two additional functions. 1) for mounting/unmounting proc file system. 2) for handling SIGTERM and sending SIGTERM to the child process (user's command)

tcsc reviewed (commented):
Just a few more nits

dustinspecker commented on rfd/0001-linux-job-worker-service.md:
```diff
@@ -0,0 +1,511 @@
+### How the server works
+
+Note: gRPC handlers each run in their own goroutine. Out of the box, this enables the server to support concurrent requests.
+The handlers must be safe for concurrent use, such as using a [sync.Map](https://pkg.go.dev/sync#Map) to store jobs.
```
You've sold me. The compiler-time type safety versus runtime type assertions is worth the trade-off of handling the mutex ourselves.

dustinspecker commented:
> Just a few more nits

All great points - thank you for taking the time to make a thorough review!
